### ISSUE TITLE
`Function` is a built-in identifier, but no error is emitted when it is used to name a type

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#45703]
Consider the following programs:

```dart
class Function {}
class C<Function> {}
void main() {}
```

```dart
typedef Function = int;
typedef F<Function> = int;
void main() {}
```

```dart
extension Function on List {}
extension E<Function> on List<Function> {}
void main() {}
```

```dart
mixin Function {}
mixin M<Function> implements List<Function> {}
void main() {}
```

The word `Function` is a [built-in identifier](https://github.com/dart-lang/language/blob/ff957765b7dd4495bdb2a74157671a0d019ef48d/specification/dartLangSpec.tex#L16025), but it is not consistently reported as an error when this word is used as the name of a type (that is, a class, type parameter, type alias, etc.).

No errors or other diagnostics are emitted except for the following:

- The analyzer hints that a class named `Function` is deprecated.
- The analyzer and the CFE emit an error "'Function' can't be used as an identifier because it's a keyword" on `typedef Function = int;`.

All declaring occurrences of `Function` should be marked as a compile-time error.

**Comment by jensjoha** [comment]
The link tries to point to line 16025 in a file with 14154 lines in it... Could you provide a better link?

**Comment by lrhn** [comment]
The link is correct, but Github sometimes doesn't show the whole file (it's 21899 lines). Try reloading.

**Comment by jensjoha** [comment]
Hah, fantastic! Thanks!

**Comment by scheglov** [comment]
There is class `Function` in `lib/core/function.dart`.
Do we make an exception for SDK?

**Comment by lrhn** [comment]
As I remember it (big caveat), we did not choose to make `Function` a built-in identifier. Instead we treated `Function` followed by `(` or `<` as a context-sensitive keyword instead of an identifier.

One of the reasons was that `Function` is a type name already, and you need to be able to use it as one, which you can't with built-in identifiers. Did we change the approach to "built-in identifier but can be used as type" instead?

**Comment by bwilkerson** [comment]
Section 17.37 Identifier Reference, as of May 6, lists `Function` as a built-in identifier. I don't know whether that's as intended.

@eernstg

**Comment by lrhn** [comment]
Nope, my bad. `Function` has apparently been a built-in identifier since Dart 2.0. Don't think I ever remembered that.
Introduced in https://dart-review.googlesource.com/c/sdk/+/41360/9/docs/language/dartLangSpec.tex

In which case the platform "class" declaration for `Function` must be special cased (just like `FutureOr`, because it's not really class, or at least it doesn't have to be).

**Comment by eernstg** [comment]
Yes, `Function` was made a built-in identifier Feb 19 2018 in cec672c7934e4c8513a58b271908efba702246a9. I don't remember the exact discussions about this change. We already have another built-in identifier which typically denotes a type: `dynamic`. So in that sense `Function` is not alone. I don't see a problem in the fact that a system library contains a `class Function ...` &mdash; that declaration is given special treatment by the analyzer/compilers anyway.

We do get an error for a type alias, for the CFE as well as the analyzer:

```dart
// Store in `n014.dart`.
typedef Function = int;
void main() {}
/*
> dart n014.dart
n014.dart:1:9: Error: 'Function' can't be used as an identifier because it's a keyword.
Try renaming this to be an identifier that isn't a keyword.
typedef Function = int;
        ^^^^^^^^
> dartanalyzer n014.dart
Analyzing n014.dart...
  error • 'Function' can't be used as an identifier because it's a keyword. • n014.dart:1:9 • expected_identifier_but_got_keyword
1 error found.
*/
```

Surprisingly, we do not get that error for a class or a mixin, but the analyzer reports that it is "deprecated" to use the name `Function` for a class. In other words, it's not fully implemented today.

So do we wish to make it a regular identifier in the spec, given that the error is not emitted for classes an mixins? I don't think it will create any serious parsing difficulties because `Function` must be handled when it occurs as a type anyway, but I doubt that it will be helpful in a software engineering sense to have a user-written class or mixin named `Function`.

**Comment by bwilkerson** [comment]
I'll just note that there is a CL to start generating diagnostics in analyzer in these cases (https://dart-review.googlesource.com/c/sdk/+/195760).

Does the fact that we use `Function` for function types, such as `int Function(String)`, impact whether it's a built-in identifier or a normal identifier?

**Comment by scheglov** [comment]
I started this in https://dart-review.googlesource.com/c/sdk/+/200080, but it seems that there are more changes to CFE than I can do on my own.

**Comment by lrhn** [comment]
> Does the fact that we use `Function` for function types, such as `int Function(String)`, impact whether it's a built-in identifier or a normal identifier?

No, when followed by `<` or `(`, the `Function` token is effectively a contextual keyword instead of an identifier. It's not resolved against the environment at all.

**Comment by leafpetersen** [comment]
I see that we issue a hint on this now, but we also seem to issue the hint for the version that is in the SDK core libraries (which, from conversation above, is supposed to be exempt?).

**Comment by bwilkerson** [comment]
Yes, we're producing a hint for class declarations (`HintCode.DEPRECATED_FUNCTION_CLASS_DECLARATION`), but we ought to be producing one of the `CompileTimeErrorCode BUILT_IN_IDENTIFIER_AS_*` codes in order to catch its use in other places. For example, I believe we should produce a diagnostic for the following, but we don't:
```
extension Function on int {}
```

@stereotype441 In case you'd be interested in fixing this, given that it spans both analyzer and CFE.

**Comment by jensjoha** [comment]
This should be fixed with 25dcfca3c23e5ca6870b4c3fc614750377d8efa1.
